### ISSUE TITLE
bug: false positive in `no-pass-data-to-parent` (inside cleanup callback)

### ISSUE THREAD

**Issue by kachkaev** [NickvanDyke/eslint-plugin-react-you-might-not-need-an-effect#37]
### Did you check docs and existing issues?

- [x] I have read the plugin docs
- [x] I have searched the existing issues
- [x] I have read [You Might Not Need An Effect](https://react.dev/learn/you-might-not-need-an-effect) and am reasonably confident that I *do* need an effect

### Plugin version

0.5.3

### ESLint version

9.36.0

### Describe the bug

I use [Pragmatic drag and drop](https://atlassian.design/components/pragmatic-drag-and-drop/about) which adds drag and drop behavior to elements inside `useEffect`. When I wire D&D callbacks to component props, I see `no-pass-data-to-parent` from this plugin. These props are not called inside `useEffect` itself but rather in event callbacks that are emitted later.

### Steps To Reproduce

```tsx
import { dropTargetForElements } from "@atlaskit/pragmatic-drag-and-drop/element/adapter";

function DeleteDropTarget({ onDelete }) {
  const ref = React.useRef<HTMLDivElement | null>(null);

  React.useEffect(() => {
    const el = ref.current;
    if (!el) {
      return;
    }

    const cleanup = dropTargetForElements({
      element: el,
      onDrop: ({ source }) => {
        onDelete(convertDragDataToOnDeletePayload(source.data)); // ← triggers no-pass-data-to-parent
      },
    });

    return cleanup;
  }, [onDelete]);

  return <div ref={ref}>Drop an item here to delete</div>;
};
```

### Expected Behavior

Calling a callback from a prop should not trigger `no-pass-data-to-parent` if it's not called _directly_ inside `useEffect` (or its cleanup function). It's impossible to pass `source.data` because it comes from the drop effect, i.e. depends on which element has been dragged and dropped above our component.

**Comment by NickvanDyke** [comment]
Yeah the rule used to ignore nested function calls but that caused a false negative on common errors like making an API request and passing the result in a Promise chain.

> onDrop would have to eventually invoke some callback from a prop or a hook

Calling a hook would not flag though, so the ref just needs to be hoisted up to the component that actually does something with this data, if I understand the library correctly. While that's more React-y, I understand it's debatable whether that's more modular/readable.

There may not be a perfect solution here. I'd guess this is a rare use, so I may prefer to disable the rule for this line than neuter the overall rule, introducing comparatively numerous false negatives.

**Comment by kachkaev** [comment]
Yeah I agree. Not sure if this edge case is actionable, at least I don’t see a good solution at this very point. Feel free to close this issue if you want or maybe wait for some more thoughts first. Up to you. Great plugin!

**Comment by NickvanDyke** [comment]
I'll think on it a bit longer! Even if unactionable, I appreciate you opening this issue to leave a trail for others 🙂
